### PR TITLE
Re-implementation of gitignore exclusions for embedding 

### PR DESCRIPTION
**Pull request by vkx86**
For my use cases, existing gitignore file exclusions  functionality didn't work good resulting in ingesting binary and other non-text files and hanging doc embedding process, so I'd re-implemented the functionality with `ignore` library. 

Me thinks that `FILE_IGNORE_LIST` should be accessible for easy user manipulation too.

### CODE REVIEW

File: src/extension/embeddings.ts [hunk 1]
```diff
@@ -13,17 +12,17 @@ import {
   Embedding,
   LMStudioEmbedding,
   RequestOptionsOllama,
-  StreamRequestOptions as RequestOptions
-} from "../common/types"
+  StreamRequestOptions as RequestOptions,
+} from '../common/types'
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:16]: Should be double.

File: src/extension/embeddings.ts [hunk 2]
```diff
@@ -13,17 +12,17 @@ import {
...
 
 import { fetchEmbedding } from "./api"
 import { TwinnyProvider } from "./provider-manager"
 import {
   getDocumentSplitChunks,
   getIgnoreDirectory,
-  readGitIgnoreFile,
   readGitSubmodulesFile
 } from "./utils"
+import ignore from "ignore"
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:25]: External libraries should be imported before local imports.

File: src/extension/embeddings.ts [hunk 3]
```diff
@@ -106,23 +110,15 @@ export class EmbeddingDatabase {
...
-            minimatch(relativePath, pattern, { dot: true, matchBase: true }) &&
-            !pattern.startsWith("!")
-          if (isIgnored) {
-            logger.log(`Ignoring ${relativePath} due to pattern: ${pattern}`)
-          }
-          return isIgnored
-        })
-      ) {
+      if(gitIgnorer.ignores(relativePath)){
+        logger.log("git-ignored: " + relativePath)
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:114]: We could use template literal instead.

File: src/extension/embeddings.ts [hunk 4]
```diff
@@ -106,23 +110,15 @@ export class EmbeddingDatabase {
...
         continue
-      }
+      }
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:116]:
```suggestion
      }
```

File: src/extension/embeddings.ts [hunk 5]
```diff
@@ -106,23 +110,15 @@ export class EmbeddingDatabase {
...
+      if(gitIgnorer.ignores(relativePath)){
+        logger.log("git-ignored: " + relativePath)
         continue
-      }
+      }   
 
       if (dirent.isDirectory()) {
         filePaths = filePaths.concat(await this.getAllFilePaths(fullPath))
       } else if (dirent.isFile()) {
+        // logger.log("add 4 embd: " + relativePath)
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:121]: Please remove comment.

File: src/extension/embeddings.ts [hunk 6]
```diff
@@ -91,11 +90,16 @@ export class EmbeddingDatabase {
   private getAllFilePaths = async (dirPath: string): Promise<string[]> => {
     let filePaths: string[] = []
     const dirents = await fs.promises.readdir(dirPath, { withFileTypes: true })
-    const gitIgnoredFiles = readGitIgnoreFile() || []
     const submodules = readGitSubmodulesFile()
 
     const rootPath = vscode.workspace.workspaceFolders?.[0]?.uri.fsPath || ""
 
+    const gitIgnorer = ignore()
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:97]: Could we call it `ig` like in the documentation?

File: .gitignore [hunk 7]
```diff
@@ -3,3 +3,4 @@ dist
 node_modules
 .vscode-test/
 *.vsix
+package-lock.json
```
[the remark this answers is not in the document]
Comment (vkx86) [reply .gitignore]: I have no sentiment for auto-generated files, but as U wish ;-)

File: .gitignore [hunk 8]
```diff
@@ -2,4 +2,4 @@ out
 dist
 node_modules
 .vscode-test/
-*.vsix
+*.vsix
```
Comment (rjmacarthy) [inline .gitignore:5]:
```suggestion
*.vsix

```

File: src/extension/embeddings.ts [hunk 9]
```diff
@@ -106,17 +110,8 @@ export class EmbeddingDatabase {
...
-          const isIgnored =
-            minimatch(relativePath, pattern, { dot: true, matchBase: true }) &&
-            !pattern.startsWith("!")
-          if (isIgnored) {
-            logger.log(`Ignoring ${relativePath} due to pattern: ${pattern}`)
-          }
-          return isIgnored
-        })
-      ) {
+      if(ig.ignores(relativePath)){
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:113]:
```suggestion
      if (ig.ignores(relativePath)){
```

File: src/extension/embeddings.ts [hunk 10]
```diff
@@ -106,17 +110,8 @@ export class EmbeddingDatabase {
...
-          const isIgnored =
-            minimatch(relativePath, pattern, { dot: true, matchBase: true }) &&
-            !pattern.startsWith("!")
-          if (isIgnored) {
-            logger.log(`Ignoring ${relativePath} due to pattern: ${pattern}`)
-          }
-          return isIgnored
-        })
-      ) {
+      if (ig.ignores(relativePath)){
```
Comment (rjmacarthy) [inline src/extension/embeddings.ts:113]:
```suggestion
      if (ig.ignores(relativePath)) {
```

**Review by rjmacarthy — APPROVED** [review]

**Comment by rjmacarthy** [comment]
Hey, due to changes in the package-lock.json file this PR broke twinny on all architectures, therefore I have had to revert the change for now, if you could help resolve the issue whilst making sure the package-lock json file remains intact without breaking changes we can re-merge.

Many thanks!

**Comment by vkx86** [comment]
How auto-generated package-lock.json can remain intact if I added dependency to `ignore` library?

**Pull request merged**